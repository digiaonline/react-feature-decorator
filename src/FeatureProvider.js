// @flow

import {Component, Children} from 'react'
import featuresShape from './featuresShape'
import type {FeatureName} from './types'

type Props = {
  features: Array<FeatureName>,
  enabledFeatures: Array<FeatureName>,
  children: any,
}

class FeatureProvider extends Component<Props, void> {
  static defaultProps = {
    features: [],
    enabledFeatures: [],
  }

  static childContextTypes = {
    features: featuresShape.isRequired,
  }

  getChildContext() {
    return {
      features: {
        isDefined: this.isDefined,
        isEnabled: this.isEnabled,
        tryIsDefined: this.tryIsDefined,
      },
    }
  }

  isDefined = (featureName: FeatureName) => {
    return this.props.features.indexOf(featureName) !== -1
  }

  tryIsDefined = (featureName: FeatureName) => {
    if (!this.isDefined(featureName)) {
      throw `Feature '${featureName}' is not defined!`
    }
  }

  isEnabled = (featureName: FeatureName) => {
    this.tryIsDefined(featureName)

    return this.props.enabledFeatures.indexOf(featureName) !== -1
  }

  render() {
    return this.props.children ? Children.only(this.props.children) : null
  }
}

export default FeatureProvider
