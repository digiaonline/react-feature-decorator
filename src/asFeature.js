// @flow

import React, {Component} from 'react'
import withFeatures from './withFeatures'

export default (featureName: string) => (WrappedComponent: any) => {
  class AsFeatureDecorator extends Component<*, *> {
    props: Object

    static WrappedComponent: any

    componentWillMount() {
      this.props.features.tryIsDefined(featureName)
    }

    render() {
      return this.props.features.isEnabled(featureName) ? (
        <WrappedComponent {...this.props} />
      ) : null
    }
  }

  AsFeatureDecorator.WrappedComponent = WrappedComponent

  return withFeatures(AsFeatureDecorator)
}
