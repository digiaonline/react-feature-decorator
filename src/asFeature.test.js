import React from 'react'
import renderer from 'react-test-renderer'
import FeatureProvider from './FeatureProvider'
import asFeature from './asFeature'

const createProvider = props => renderer.create(<FeatureProvider {...props} />)

test('Enabled features are rendered', () => {
  const EnabledFeature = asFeature('myFeature')(() => (
    <div>This is my enabled feature</div>
  ))

  const tree = createProvider({
    features: ['myFeature'],
    enabledFeatures: ['myFeature'],
    children: <EnabledFeature />,
  })

  expect(tree.toJSON()).toMatchSnapshot()
})

test('Disabled features are not rendered', () => {
  const DisabledFeature = asFeature('myFeature')(() => (
    <div>This is my disabled feature</div>
  ))

  const tree = createProvider({
    features: ['myFeature'],
    children: <DisabledFeature />,
  })

  expect(tree.toJSON()).toMatchSnapshot()
})

test('Undefined features throws an error', () => {
  const UndefinedFeature = asFeature('myFeature')(() => (
    <div>This is my undefined feature</div>
  ))

  expect(() => createProvider({children: <UndefinedFeature />})).toThrow()
})
