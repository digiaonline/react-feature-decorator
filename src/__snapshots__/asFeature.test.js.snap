// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Disabled features are not rendered 1`] = `null`;

exports[`Enabled features are rendered 1`] = `
<div>
  This is my enabled feature
</div>
`;
